src/USER-HSMA/hsma2d.cpp

```diff
@@ -0,0 +1,2056 @@
+		
+		int lx = ceil((Rs - Lx / 2.0) / Lx), ly = ceil((Rs - Ly / 2.0) / Ly), lz = ceil((Rs - Lz / 2.0) / Lz);
+		int TotalNumber = nlocal * (2 * lx + 1) * (2 * ly + 1) * (2 * lz + 1);
+		double ImageCharge[TotalNumber][5];
+		int ImageNumber;
+		SetImageCharge(ImageCharge, &ImageNumber, (2 * lx + 1)* (2 * ly + 1)* (2 * lz + 1), X, Q, nlocal, Rs, Lx, Ly, Lz, lx, ly, lz);
+
+		/*Calculate Near Field and ZD*/
+		CalculateNearFieldAndZD((double**)TopNear, (double**)TopZDNear, (double**)DownNear, (double**)DownZDNear, ImageCharge, ImageNumber, (double***)IntegralTop, (double***)IntegralDown, Nw, Gamma, IF_FMM_RightTerm, S, AR, Lx, Ly, Lz, tolerance);
+		
+		/*Construct the right term by using the DTN condition*/
+		double RightTermReal[2 * NJK], RightTermImag[2 * NJK];
+		ConstructRightTerm(RightTermReal, RightTermImag, (double**)TopNear, (double**)TopZDNear, (double**)DownNear, (double**)DownZDNear, (double***)IntegralTop, (double***)IntegralDown, Nw, NJKBound, EU, mu, Lx, Ly, Lz, S, tolerance);
+		double RightTermReal_All[2 * NJK], RightTermImag_All[2 * NJK];
+		MPI_Reduce(RightTermReal, RightTermReal_All, 2 * NJK, MPI_DOUBLE, MPI_SUM, 0, world);
+		MPI_Reduce(RightTermImag, RightTermImag_All, 2 * NJK, MPI_DOUBLE, MPI_SUM, 0, world);
+
+		//Solve The Least Square Problem
+		double C[p * p];
+		if (me == 0) {
+			SolveLeastSquareProblem(C, (double**)LeftTermReal, (double**)LeftTermImag, RightTermReal, RightTermImag, p, 2 * NJK);
+	    }
+		MPI_Bcast(C, p* p, MPI_DOUBLE, 0, world);
+		MPI_Wait(&request, &status);
+		MPI_Wait(&request_Q, &status);
+
+		double Energy_HSMA;
+		double Force[nlocal][3], Pot[nlocal];
+		TotalNumber = ceil(maxatom * (2 * lx + 1) * (2 * ly + 1) * (2 * lz + 1) / 2);//¼õÉÙ´æ´¢Á¿ÏûºÄ
+		double ImageCharge_All[TotalNumber][5];
+		SetImageCharge(ImageCharge_All, &ImageNumber, (2 * lx + 1)* (2 * ly + 1)* (2 * lz + 1), AllSource, AllQ, maxatom, Rs, Lx, Ly, Lz, lx, ly, lz);
+		
+		/*Calculate the potential, energyand force of the source charges*/
+		Energy_HSMA = FinalCalculateEnergyAndForce(Force, Pot, X, Q, nlocal, ImageCharge_All, ImageNumber, Fibonacci, (double**)QRD, (double**)QLocalRD, Gamma,C, p, Fp, F, Rs, PI, IF_FMM_FinalPotential, tolerance);
+		
+		scale = 1.0;
+		const double qscale = qqrd2e * scale;
+
+		//Assemble results
+		for (int i = 0; i < nlocal; i++)
+		{
+			f[i][0] += Force[i][0] * qscale;
+			f[i][1] += Force[i][1] * qscale;
+			f[i][2] += Force[i][2] * qscale;
+		}
+
+		double Energy_HSMA_All;
+		MPI_Allreduce(&Energy_HSMA, &Energy_HSMA_All, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+		if (eflag_global) {
+			energy += Energy_HSMA_All * qscale / 2;
+		}
+		if (evflag_atom) {
+			if (eflag_atom) {
+				for (int i = 0; i < nlocal; i++) {
+					eatom[i] += Pot[i];
+					eatom[i] *= qscale;
+				}
+			}
+
+			if (vflag_atom)
+				for (int i = 0; i < nlocal; i++)
+					for (int j = 0; j < 6; j++) vatom[i][j] *= q[i] * qscale;
+		}
+	}
+
+	time = MPI_Wtime() - time;
```

Comment:
Please note that "time" is a function in the C library (thus should be avoided as a variable name to avoid confusion). Also tallying of time for KSpace will already be done by the Verlet or Respa classes.

Reply:
Removed